[COLLECTIONS-843] Implement Layered Bloom filter

Claudenw opened the pull request:
Implementation to fix [COLLECTIONS-843](https://issues.apache.org/jira/browse/COLLECTIONS-843).

This is a simple Layered Bloom filter implementation with complex configuration options.

A static method to create a fixed level implementation that does not merge is provided.
An example of a complex time based Bloom filter is provided in the testing code where filters are added to until they are saturated or one second has elapsed and are removed after they are 4 seconds old.

this change is in support of Kafka KIP-936 
https://cwiki.apache.org/confluence/display/KAFKA/KIP-936%3A+Throttle+number+of+active+PIDs

Claudenw committed:
Adjusted tests to handle bloom filter implementations that utilized
automatic decay.

Claudenw committed:
fixed indent

Claudenw committed:
updated for layered testing

Claudenw committed:
removed spaces

Claudenw committed:
cleaned up spacing

Claudenw committed:
fixed merge issue

Claudenw committed:
initial checkin

Claudenw committed:
cleaned up tests

Claudenw committed:
fixed timing on test

Claudenw committed:
fixed formatting

Claudenw committed:
added javadoc

Claudenw committed:
fixed typos

Claudenw committed:
removed blank lines

Claudenw committed:
fixed javadocs

garydgregory committed:
Fix Javadoc

garydgregory committed:
Add Javadoc  @since 4.5

garydgregory committed:
Add Javadoc  @since 4.5

Claudenw committed:
updated tests and added BloomFilterProducer code

Claudenw commented:
> 

mvn doesn't catch all the issues :(  I just spend a number of hours working out new tests for coverage changes.

Claudenw committed:
Merge branch 'apache:master' into layered_filter

Claudenw committed:
Merge branch 'layered_filter' of github.com:Claudenw/commons-collections into layered_filter

Claudenw committed:
Cleaned up javadoc and BiPredicate<BloomFilter,BloomFilter> processing

Claudenw committed:
fixed javadoc issues

Claudenw committed:
fixed typography issue

Claudenw committed:
Fixed a documentation error

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilterTest.java:
```diff
@@ -0,0 +1,325 @@
+        @Override
+        public boolean test(LayerManager lm) {
+            int depth = lm.getDepth();
+            if (depth == 0) {
```
Can this be zero? Even a clear operation then adds back a single filter layer.

aherbert reviewed (changes requested):
Hi @Claudenw,

There is a lot to review here. The main questions are why we allow access in the LayerManager and LayeredBloomFilter to the filter layers. This is used in testing (fair enough) but does it need to be public? I think it dangerous to allow a user direct access to get and clear layers.

Note that this cannot be totally prevented as you can do this anyway using the `forEachBloomFilter` method to get at the individual layers by reference. But exposing the get/clear functionality in the API seems to not be required for the filter to work.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/TestingHashers.java:
```diff
@@ -88,4 +90,18 @@ public static <T extends BloomFilter> T populateRange(T filter, int start, int e
+    /**
+     * Creates an EnhancedDoubleHasher hasher from 2 random longs.
+     */
+    public static Hasher randomHasher() {
```
True,  but when developing new Bloom filter implementations (not necessarily in this package) having the randomHasher implementation handy is a great help.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/WrappedBloomFilter.java:
```diff
@@ -0,0 +1,143 @@
+ *
+ * @since 4.5
+ */
+public abstract class WrappedBloomFilter implements BloomFilter {
```
Let's consider this for a moment.  The example in the test code is precisely what the Kafka KIP-936 is looking for.  Now, I don't want to implement a "strange" bloom filter in this class, but I do want an easy way to extend the classes we have.  Since they are all final decorating them with a wrapper is the only reasonable solution.  So let's provide the solution.  If we move it out, then I will have to ask to move it back when the KIP-936 begins development using this library.  I believe they are starting that process now.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayerManager.java:
```diff
@@ -0,0 +1,349 @@
+         * Advances the target once a merge has been performed.
+         */
+        public static final Predicate<LayerManager> ADVANCE_ON_POPULATED = lm -> {
+            return !lm.filters.isEmpty() && !lm.filters.peekLast().forEachBitMap(y -> y == 0);
```
I agree.  I will add that.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    @Override
+    public boolean forEachIndex(IntPredicate predicate) {
```
The notes for the method indicate that index ordering and uniqueness are not guaranteed.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    @Override
+    public int estimateIntersection(final BloomFilter other) {
```
The Bloom filter class has always said that operations across Bloom filters are only valid it the Shape is the same.  We do not verify that on every operation as it would slow the operations down significantly, but we do provide shape so users can make the check if they desire.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+     *
+     * @see LayerManager
+     */
+    public void next() {
```
If you have a LayeredBloomFilter that never advances then you can use some external trigger for force it to advance.

Claudenw committed:
code format cleanup

Claudenw committed:
code simplification and documentation

Claudenw committed:
added isEmpty and associated tests

Claudenw committed:
Changes as requested by review

Claudenw committed:
cleaned up formatting errors

Claudenw committed:
fixed javadoc issues

Claudenw committed:
added LayeredBloomFilter to overview.

Claudenw committed:
added coco driven test cases.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+
+    @Override
+    public int characteristics() {
+        return 0;
```
The layered Bloom filter makes some assumptions that work best when it is considered non-sparse. As we don't know what other characteristics may be added in the future, it seems that defining the characteristics here is correct.  In addition there is no requirement that all the layers be of the same characteristic.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    @Override
+    public boolean forEachBitMap(LongPredicate predicate) {
```
By definition forEachBitMap may only return the number of BitMaps specified by the Shape or fewer, never more.

Claudenw committed:
attempt to fix formatting

Claudenw commented:
I have made a mess of the formatting.  I am working at cleaning it up now.

Claudenw committed:
cleaned up javadoc differences

Claudenw committed:
cleaned up javadoc

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/Shape.java:
```diff
@@ -227,6 +227,15 @@ public double estimateN(final int cardinality) {
     }
 
+    /**
+     * Estimates the maximum number of elements that can be merged into a filter of this shape before
```
text has been changed to read:

Estimates the maximum number of elements that can be merged into a filter of
     * this shape before the false positive rate exceeds the desired rate. <p> The
     * formula for deriving {@code k} when {@code m} and {@code n} are known is:
     *
     * <pre>
     * k = ln2 * m / n
     * </pre>
     *
     * solving for {@code n} yields:
     *
     * <pre>
     * n = ln2 * m / k
     * </pre>

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/WrappedBloomFilter.java:
```diff
@@ -0,0 +1,143 @@
+public abstract class WrappedBloomFilter implements BloomFilter {
+    final BloomFilter wrapped;
+
+    public WrappedBloomFilter(BloomFilter bf) {
```
done

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    @Test
+    public void testAdvanceOnSaturationWithInt() {
```
Removed int version

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilterTest.java:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    @Test
+    public void testExpiration() throws InterruptedException {
```
changed as per your comments.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/TestingHashers.java:
```diff
@@ -88,4 +90,18 @@ public static <T extends BloomFilter> T populateRange(T filter, int start, int e
         return filter;
     }
+
+    private static Random random;
```
switched to ThreadLocalRandom.current()

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+     * @return the Bloom filter at the specified depth.
+     * @throws NoSuchElementException if depth is not in the range [0,getDepth())
+     */
+    public BloomFilter get(int depth) {
```
Because there can be additional info associated with the Bloom filter.  In the paper and in many implementations the find returns the deepest node that matches.  Our finds returns the indexes that match so that the application may select the deepest, the shallowest, the one from the middle, or perhaps all of them and do some subsequent processing.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilter.java:
```diff
@@ -0,0 +1,416 @@
+     *
+     * @return the merged bloom filter.
+     */
+    public BloomFilter flatten() {
```
I updated the comment to reflect that an empty filter should be returned if there are no filters in the collection.

I don;t have a problem adding flatten() to the BloomFilterProducer.  I will do that shortly.

Claudenw replied:
added flatten() to BloomFilterProducer.

Claudenw committed:
Made flatten() part of BloomFilterProducer

Claudenw committed:
fixed since tag.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayerManager.java:
```diff
@@ -0,0 +1,349 @@
+     * Static methods an variable for standard extend checks.
+     *
+     */
+    public static class ExtendCheck {
```
Done

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayerManager.java:
```diff
@@ -0,0 +1,349 @@
+         * @param breakAt the number of filters to merge into each filter in the list.
+         * @return A Predicate suitable for the LayerManager extendCheck parameter.
+         */
+        public static Predicate<LayerManager> advanceOnCount(int breakAt) {
```
done

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/LayerManager.java:
```diff
@@ -0,0 +1,349 @@
+        }
+
+        public LayerManager build() {
+            if (supplier == null) {
```
done

Claudenw committed:
changed X() methods to setX()

Claudenw committed:
updated javadoc

Claudenw commented:
Recent changes:
# Main code
## BloomFilterProducer

* Cleaned up documentation to specify that implementations should specify if filters used in `forEachBloomFilterPair()` and `asBloomFilterArray` are copies or references.
* Layer and depth order discussion moved to `LayerManager`.
* `asBloomFilterArray` coding change accepted, now uses List internally and returns array.
* `forEachBloomFilterPair` documentation corrected.
* added `flatten()`

## CountingPredicate

* documentation updated

## LayerManager

* documentation updated.
* `ExtendCheck` and `Cleanup` classes changed to have only static methods, no public static variables.
* added parameter checks for `ExtendCheck.advanceOnCount#breakAt`, `ExtendCheck.advanceOnSaturation#maxN`, and `Cleanup.onMaxSize#maxSize`
* removed the ``ExtendCheck.advanceOnSaturation(int)` method
* LinkedList is required as the Bloom filter storage implementation (and `cleanup` argument) as we need rapid access to the intermediate layers in order to return the specific layers.
* added method `addFilter()` to verify that the `filterSupplier` did not return `null`.
* optimised `copy()` so that no filters are created during the construction
* moved the check for empty filter in `next()` to be part of the cleanup consumer.
* the `get()` method needs to be public so that access to the specific layer is permitted as required by the `LayeredBloomFilter`.
* `target()` renamed to `getTarget()`
* `clear()` now calls `addFilter()` to ensure a non-null filter is added.
* `Builder.withX` pattern change to `Builder.setX`
* Builder method documentation added.
* changed Builder exceptions from IllegalArgument to NPE
* The quesiton of whether to use a Builder here is still open.

## BloomFilter

* added `isEmpty()`

## LayeredBloomFilter

* Changed the hyphens in the cite for the article but left the "V1"s in the page numbers as that is how the publisher asks for it to be cited.
* updated documentation
* Access to layers is provided because there can be additional info associated with the Bloom filter. In the paper and in many implementations the find returns the deepest node that matches. Our finds returns the indexes that match so that the application may select the deepest, the shallowest, the one from the middle, or perhaps all of them and do some subsequent processing.
* removed `clear(int)` and `target()`
* added `contains(BloomFilterProducer)` to handle the special case.
* Continued to return 0 (zero) for the `characteristics()` as the layered Bloom filter makes some assumptions that work best when it is considered non-sparse. As we don't know what other characteristics may be added in the future, it seems that defining the characteristics here is correct. In addition there is no requirement that all the layers be of the same characteristic.
* notes on `forEachIndex()` in the general case notes for the method indicate that index ordering and uniqueness are not guaranteed.
* `forEachBitMap()` may only return the nbumber of BitMaps specified by the Shape so it must use `flatten()`.
* simplified `forEachBitMap()`, `estimateN()`, and `estimateUnion()` to explicitly use `flatten()`
* removed redundant `estimateUnion()`
* the use case for `next()` is using `ExtendCheck.neverAdvance()` and then triggering the advance on some outside state change/trigger.

## Shape

* clarified `estimageMaxN()` documentation
* fixed formatting

## WrappedBloomFilter

* argument for not moving code to test directory:  Let's consider this for a moment. The example in the test code is precisely what the Kafka KIP-936 is looking for. Now, I don't want to implement a "strange" bloom filter in this class, but I do want an easy way to extend the classes we have. Since they are all final decorating them with a wrapper is the only reasonable solution. So let's provide the solution. If we move it out, then I will have to ask to move it back when the KIP-936 begins development using this library. I believe they are starting that process now.
* Updated documentation

# Testing code

## DefaultBloomFilterTest

* Provided access to the two (2) BloomFilter test classes that implement the Defaults (Sparse and non-Sparse).


## LayerManagerTest

* Significantly modified `testOnMaxSize()` and `testAdvanceOnCount` to test wider range of values, including breaking out parameterized tests.
* Adjusted to match the methods now provided in LayerManager

## LayeredBloomFilterTest

* Removed checks of zero depth as it can not be returned.
* Reduced time in `testExpiration()` to approx 600ms

## TestingHashers

* converted to `ThreadLocalRandom.current()`
* left `randomHasher()` as public so that it can be utilized in 3rd party Bloom filter testing.

Claudenw committed:
fixed javadoc errors

Claudenw committed:
Merge branch 'master' into layered_filter

Claudenw committed:
merged changes from master

Claudenw committed:
renamed to Test to CellProducerFromLayeredBloomFilterTest

Claudenw commented:
@garydgregory @aherbert  I have merged the latest main branch into this pull request and the last major changes are summarized above.

Claudenw committed:
changed to jupiter from junit.

Claudenw committed:
added override for uniqueIndices as optimization.

Claudenw committed:
fixed checkstyle issue

Claudenw commented:
@garydgregory @aherbert 

Gentlemen, 

Can I get eyes on the new code please.  I think it is correct and fixes all the previously identified issues.

Claudenw commented:
@garydgregory @aherbert nudge

garydgregory reviewed (approved):
OK here. @aherbert ?

aherbert commented on src/main/java/org/apache/commons/collections4/bloomfilter/Shape.java:
```diff
@@ -228,12 +228,24 @@ public double estimateN(final int cardinality) {
+     * k = ln2 * m / n
+     * </pre>
+     *
+     * solving for {@code n} yields:
```
Should have a `<p>`

aherbert reviewed (changes requested):
Noted a few changes. I did not look at the test code. I will do so in the next few days.

Claudenw committed:
modified as per review

Claudenw commented:
@aherbert thank you for your review.  Changes have been made to the base code.  One of these days I will figure out how you see all those things that I seem to miss.

Claudenw committed:
Merge branch 'layered_filter' of github.com:Claudenw/commons-collections into layered_filter

garydgregory commented:
Let's make sure we let GitHub squash the commits once this is approved.

Claudenw commented:
@aherbert Have you had a chance to look at the test code?  I have a series of blog posts about Bloom filters coming out "soon" (depending on when my company gets around to finalizing it).  The second one in the series talks about commons-collections Bloom filters.  I am hoping we have this in the 4.5-SNAPSHOT before then.

Thanks to you and @garydgregory  for all your assistance in making the entire package a better submission.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    @Test
+    public void testAdvanceOnSaturation() {
```
This is exactly the same as `testAdvanceOnCalculatedFull`. I think this should be:
```java
double maxN = shape.estimateMaxN();
Predicate<LayerManager> underTest = LayerManager.ExtendCheck.advanceOnSaturation(maxN);
```

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,305 @@
+        LayerManager layerManager = testingBuilder().build();
+        while (layerManager.getTarget().getShape().estimateN(layerManager.getTarget().cardinality()) < maxN) {
+            assertFalse(underTest.test(layerManager));
+            layerManager.getTarget().merge(TestingHashers.randomHasher());
```
I was concerned by the use of random hasher. However with the shape as (17, 72) I observe only 3 or 4 hashers required to saturate the shape. Maybe add this as a note with the maxN value (2.9356) in case the shape ever changes in the future so it is obvious how to update the test.

Claudenw replied:
If the shape changes the loop will execute until the maxN is reached.  the estimatedN will always reach maxN before the all the bits are enabled.  So I am not certain what the issue is here.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,305 @@
+        LayerManager layerManager = testingBuilder().build();
+        while (layerManager.getTarget().getShape().estimateN(layerManager.getTarget().cardinality()) < maxN) {
+            assertFalse(underTest.test(layerManager));
+            layerManager.getTarget().merge(TestingHashers.randomHasher());
```
When I use maxN = number of bits I see the use of a random hasher requires 11-30 iterations (sample size = 20). So this is fine but a larger shape could be an issue to fill up. Could we replace this with a series of `IncrementingHasher` starting from 1 and increasing the initial value by (number of hash functions) each loop iteration?

Claudenw replied:
done.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,305 @@
+        assertThrows(IllegalArgumentException.class, () -> LayerManager.ExtendCheck.advanceOnSaturation(-1));
+    }
+
+    private void testOnMaxSize(int maxSize) {
```
Should be a ParameterizedTest.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,305 @@
+
+    @Test
+    public void testGet() {
+        LayerManager underTest = LayerManager.builder().setSupplier(() -> new SimpleBloomFilter(shape)).build();
```
The SimpleBloomFilter could be cached and then returned. You can then assert that the LayerManager either does or does not return a copy:
```java
SimpleBloomFilter f = new SimpleBloomFilter(shape);
LayerManager underTest = LayerManager.builder().setSupplier(() -> f).build()
// ...
assertSame(f, underTest.get(0));
```

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilterTest.java:
```diff
@@ -0,0 +1,321 @@
+    public void testFindBloomFilter() {
+        LayeredBloomFilter filter = setupFindTest();
+        int[] result = filter.find(TestingHashers.FROM1);
+        assertEquals(2, result.length);
```
Should use assertArrayEquals to directly test the layers in which the hasher is found

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilterTest.java:
```diff
@@ -0,0 +1,321 @@
+        BitMapProducer producer = BitMapProducer.fromIndexProducer(idxProducer, getTestShape().getNumberOfBits());
+
+        int[] result = filter.find(producer);
+        assertEquals(2, result.length);
```
Use assertArrayEquals

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayeredBloomFilterTest.java:
```diff
@@ -0,0 +1,321 @@
+        LayeredBloomFilter filter = setupFindTest();
+
+        int[] result = filter.find(producer);
+        assertEquals(2, result.length);
```
Use assertArrayEquals

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilterTest.java:
```diff
@@ -41,4 +41,16 @@ public void testMergeShortBitMapProducer() {
     }
+
+    @Test
+    public void testCardinalityAndIsEmpty() {
```
This is just doing what the parent class does. The extra test should be in its own fixture, e.g.
```java
@Test
public void testEmptyAfterMergeWithNothing() {
    // test the case where is empty after merge
    // in this case the internal cardinality == -1
    BloomFilter bf = createEmptyFilter(getTestShape());
    bf.merge(IndexProducer.fromIndexArray());
    assertTrue(bf.isEmpty());
}
```

Claudenw replied:
Added as fixture in abstract test class.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/WrappedBloomFilterTest.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Test
+    public void testCharacteristics() {
```
I do not know why but the characteristics method had no coverage when I ran tests locally. I changed this to:
```java
@ParameterizedTest
@ValueSource(ints = {0, 1, 34})
public void testCharacteristics(int characteristics) {
    Shape shape = getTestShape();
    BloomFilter inner = new DefaultBloomFilterTest.SparseDefaultBloomFilter(shape) {
        @Override
        public int characteristics() {
            return characteristics;
        }
    };
    WrappedBloomFilter underTest = new WrappedBloomFilter(inner) {};
    assertEquals(characteristics, underTest.characteristics());
}
```
Now I see coverage.

Claudenw replied:
fixed

aherbert reviewed (changes requested):
@Claudenw There are some minor things to clean up here but no major issues. Can you check the LayerManagerTest for testAdvanceOnCalculatedFull and testAdvanceOnSaturation are using the correct `maxN`.

garydgregory commented:
Hi All. Is this PR ready to be merged?

aherbert commented:
There are some changes to make to the test code. I only reviewed this yesterday.

garydgregory commented:
> There are some changes to make to the test code. I only reviewed this yesterday.

👍  TY for the update

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/DefaultBloomFilterTest.java:
```diff
@@ -215,9 +215,12 @@ public int cardinality() {
+    public static class SparseDefaultBloomFilter extends AbstractDefaultBloomFilter {
 
-        SparseDefaultBloomFilter(final Shape shape) {
+        public SparseDefaultBloomFilter(final Shape shape) {
```
Yes, that is exactly the usage.  In testing a new BloomFilter implementation is is handy to have a clean implementation of a sparse filter as used in the testing to verify correct functioning of the new Bloom filter.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/DefaultBloomFilterTest.java:
```diff
@@ -234,9 +237,12 @@ public AbstractDefaultBloomFilter copy() {
+    /**
+     * A default implementation of a non-sparse Bloom filter.
+     */
+    public static class NonSparseDefaultBloomFilter extends AbstractDefaultBloomFilter {
```
same as above

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java:
```diff
@@ -0,0 +1,305 @@
+    @Test
+    public void testAdvanceOnCalculatedFull() {
+        Double maxN = shape.estimateMaxN();
+        Predicate<LayerManager> underTest = LayerManager.ExtendCheck.advanceOnSaturation(shape.estimateMaxN());
```
This test is being removed it is a holdover from when there were 2 methods: advanceOnCalculatedFull() and advanceOnSaturation()

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/CountingPredicateTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertTrue(cp.forEachRemaining());
+
+        // test last item not checked
+        cp = new CountingPredicate<>(ary, (x, y) -> y == Integer.valueOf(2));
```
done

Claudenw committed:
Updated tests as per review

Claudenw committed:
fixed variable initialization issues

Claudenw committed:
Merge branch 'master' into layered_filter

Claudenw commented:
@aherbert just a friendly ping.  I'd like to get this closed.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/CountingPredicateTest.java:
```diff
@@ -16,31 +16,103 @@
 
 public class CountingPredicateTest {
+    private List<Pair<Integer, Integer>> expected = new ArrayList<>();
+    private List<Pair<Integer, Integer>> result = new ArrayList<>();
```
This could be an argument to `makeFunc`. The `expected` can be created for each test that uses it.

Currently using a class level object works as JUnit is only executing one test at a time on an instance. But it is cleaner and more thread safe to pass in `result` to the `makeFunc` helper (which can be static).

Claudenw replied:
@aherbert,  thanks for the suggestion.  I made the changes.

aherbert reviewed (commented):
Only reviewed the changes since the last review. It looks OK. I noted a change to make the test able to support concurrent test execution.

Claudenw committed:
made suggested test changes

Claudenw committed:
fixed broken test

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/CountingPredicateTest.java:
```diff
@@ -28,11 +28,11 @@
 public class CountingPredicateTest {
-    private List<Pair<Integer, Integer>> expected = new ArrayList<>();
-    private List<Pair<Integer, Integer>> result = new ArrayList<>();
+    //private List<Pair<Integer, Integer>> expected = new ArrayList<>();
```
@Claudenw I don't think we need to keep these.

garydgregory commented:
@aherbert @Claudenw 
Is this PR ready to merge? If not, what remains? TY!

aherbert commented:
I previously requested Claude to remove some commented out code in a test class. Otherwise this is fine.

garydgregory committed:
Remove dead comments per code reviews
